private/bufnew/bufconfig/generate_managed_config.go

```diff
@@ -0,0 +1,82 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufconfig
+
+import (
+	"github.com/bufbuild/buf/private/bufnew/bufmodule"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+// GenerateManagedConfig is a managed mode configuration.
+type GenerateManagedConfig interface {
+	// Enabled returns whether managed mode is enabled.
+	Enabled() bool
+	// The first return value is the value to modify the file's option to. For
+	// example, if a config has set java_package_prefix to foo for a file with
+	// package bar, this returns "foo.bar" for the first value.
+	// The second value indicates whether this option should be modified.
+	// Reasons not to modify:
+	// 1. Some options are not supposed to be modified unless a value is specified,
+	//    such as optimize_for.
+	// 2. In the config the user has specified not to modify this option for this file.
+	// 3. The file already has the desired file option value.
+	// 4. The file is a WKT.
+	// TODO: 3 and 4 are debatable, asking the question what is the responsibility
+	// of a managed config.
+	//
+	// Note: this means a GenerateManagedConfig's interface does not deal with
+	// options like java_package_prefix and java_package_suffix, because it returns
+	// the final value to modify java_package to.
+	ValueForFileOption(imageFile, fileOption) (interface{}, bool)
```

Comment:
How am I supposed to use this outside of this package?

Reply:
```
for _, imageFile := range imageFiles {
  javaPackage, shouldModify := config.ValueForFileOption(imageFile, fileOptionJavaPackage)
  javaPackageValue, ok := javaPackage.(string)
  if !ok {
    return someErr
  }
  if !shouldModify {
    continue
  }
  imageModify.ModifyJavaPackage(marker, imageFile, javaPackageValue)
  
  // continue to the next option
}
```

or perhaps something like
```
        for _, imageFile := range imageFiles {
		for _, fileOption := range fileOptions {
			getFunc := option.GetFunc() //
			setFunc := option.SetFunc()
			value, shouldModify := config.ValueForFileOption(option)
			if !set {
				continue
			}
			if value == getFunc(imageFile) {
				continue
			}
			setFunc(marker, imageFile, value)
		}
		for _, field := range image.Fields := {
			for _, fieldOption := range fieldOptions {
			}
		}
	}
```